database/src/service/read.rs

```diff
@@ -228,57 +317,191 @@ fn map_request(
+
+    let vec = range
+        .into_par_iter()
+        .map(|block_height| {
+            let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+            get_block_extended_header_from_height(&block_height, tables)
+        })
+        .collect::<Result<Vec<ExtendedBlockHeader>, RuntimeError>>()?;
+
+    Ok(Response::BlockExtendedHeaderInRange(vec))
 }
 
 /// [`ReadRequest::ChainHeight`].
 #[inline]
-fn chain_height(env: &Arc<ConcreteEnv>) -> ResponseResult {
-    todo!()
+fn chain_height(env: &ConcreteEnv) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = env_inner.tx_ro()?;
+    let table_block_heights = env_inner.open_db_ro::<BlockHeights>(&tx_ro)?;
+    let table_block_infos = env_inner.open_db_ro::<BlockInfos>(&tx_ro)?;
+
+    let top_height = top_block_height(&table_block_heights)?;
+    let block_hash = get_block_info(&top_height, &table_block_infos)?.block_hash;
+
+    Ok(Response::ChainHeight(top_height, block_hash))
 }
 
 /// [`ReadRequest::GeneratedCoins`].
 #[inline]
-fn generated_coins(env: &Arc<ConcreteEnv>) -> ResponseResult {
-    todo!()
+fn generated_coins(env: &ConcreteEnv) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = env_inner.tx_ro()?;
+    let table_block_heights = env_inner.open_db_ro::<BlockHeights>(&tx_ro)?;
+    let table_block_infos = env_inner.open_db_ro::<BlockInfos>(&tx_ro)?;
+
+    let top_height = top_block_height(&table_block_heights)?;
+
+    Ok(Response::GeneratedCoins(cumulative_generated_coins(
+        &top_height,
+        &table_block_infos,
+    )?))
 }
 
 /// [`ReadRequest::Outputs`].
 #[inline]
-#[allow(clippy::needless_pass_by_value)] // TODO: remove me
-fn outputs(env: &Arc<ConcreteEnv>, map: HashMap<Amount, HashSet<AmountIndex>>) -> ResponseResult {
-    todo!()
+fn outputs(env: &ConcreteEnv, map: HashMap<Amount, HashSet<AmountIndex>>) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let (tx_ro, tables) = set_tx_ro_and_tables!(env, env_inner);
+
+    // -> Result<(AmountIndex, OutputOnChain), RuntimeError>
+    let inner_map = |amount, amount_index| {
+        let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+
+        let pre_rct_output_id = PreRctOutputId {
+            amount,
+            amount_index,
+        };
+        let output = get_output(&pre_rct_output_id, tables.outputs())?;
+
+        //---  Map `Output` -> `OutputOnChain`
+
+        // FIXME: implement lookup table for common values:
+        // <https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/ringct/rctOps.cpp#L322>
+        let commitment = ED25519_BASEPOINT_POINT + H() * Scalar::from(amount);
+
+        let time_lock = if output
+            .output_flags
+            .contains(OutputFlags::NON_ZERO_UNLOCK_TIME)
+        {
+            // TODO: how to recover the timelock height/time?
+            todo!()
+        } else {
+            Timelock::None
+        };
```

Comment:
Context: `ReadRequest::Outputs`, we're mapping `cuprate_database::types::Output` -> `cuprate_types::OutputOnChain`.

How do we recover this `Timelock`? I don't think we store the value directly anywhere.

We could deserialize the parent transaction from `TxBlob` then get the timelock, but that seems expensive to do for every output?

Reply:
From the `Tx Unlock Time` table: https://github.com/Cuprate/cuprate/tree/db-schema/database#tx-unlock-time

Reply:
Ahhh...